### ISSUE TITLE
In Java 8, adding annotations like @Nullable to constructors results in failed mapping.

### ISSUE THREAD

**Issue by impati** [mapstruct/mapstruct#3570]
### Expected behavior

I would like to bring to your attention a minor issue I encountered in Java 8, which I did not experience in Java 17. I am eager to understand the cause and find a solution, so I'm sharing the scenario I faced.

I have created a simple [repository](https://github.com/impati/mapstruct) to simulate the situation for your reference, if needed.

Initially, the classes MemberRequest and MemberDto exist in the api module and the domain module, respectively.

I aim to map from MemberRequest to MemberDto.

Both classes share the same field names: a String type field named name and an Integer type field named age.

```java
@Mapper
public interface MemberMapper {

    MemberDto toMemberDto(MemberRequest request);
}
```

I defined the Mapper class to convert MemberRequest to MemberDto. The expected result is as follows:

```java

@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2024-04-13T18:27:02+0900",
    comments = "version: 1.5.5.Final, compiler: IncrementalProcessingEnvironment from gradle-language-java-4.10.jar, environment: Java 1.8.0_402 (Amazon.com Inc.)"
)
@Component
public class MemberMapperImpl implements MemberMapper {

    @Override
    public MemberDto toMemberDto(MemberRequest request) {
        if ( request == null ) {
            return null;
        }

        String name = null;
        Integer age = null;

        name = request.getName();
        age = request.getAge();

        MemberDto memberDto = new MemberDto( name, age );

        return memberDto;
    }
}
```

### Actual behavior

The issue is that if the constructor of MemberDto contains annotations like @Nullable, the creation of MemberMapperImpl fails.

```java
import lombok.Getter;
import org.springframework.lang.Nullable;

@Getter
public class MemberDto {

    private final String name;
    private final Integer age;

    public MemberDto(String name, @Nullable Integer age) {
        this.name = name;
        this.age = age;
    }
}
```

```java
@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2024-04-13T18:19:20+0900",
    comments = "version: 1.5.5.Final, compiler: IncrementalProcessingEnvironment from gradle-language-java-4.10.jar, environment: Java 1.8.0_402 (Amazon.com Inc.)"
)
@Component
public class MemberMapperImpl implements MemberMapper {

    @Override
    public MemberDto toMemberDto(MemberRequest request) {
        if ( request == null ) {
            return null;
        }

        String arg0 = null;
        Integer arg1 = null;

        MemberDto memberDto = new MemberDto( arg0, arg1 );

        return memberDto;
    }
}
```

To summarize, in a mapping from class A to B, where classes A and B exist in different modules, the mapping fails if there are annotations present in the constructor of class B.

This problem did not occur when upgrading the Java version (e.g., Java 17), and I observed that using @ConstructProperties allows for proper mapping even if annotations like @Nullable are present in the constructor.

### Steps to reproduce the problem

1. Class a in Module A.
2. Class b in Module B.
3. Create a Mapper from a to b.
4. Add annotations to the constructor of class b.

### MapStruct Version

MapStruct 1.5.5

**Comment by filiphr** [comment]
This is not a bug in MapStruct. MapStruct is using the Java compiler to get the names of source parameters for classes. This means that in the scenario you are showing, when the mapper is recompiled, but the other modules aren't, then MapStruct is going to see that the parameters for the class are `arg0` and `arg1`. If you add `-parameters` to the `compilerArgs` then the byte code is going to contain the name of the parameters and MapStruct will do the mapping properly.
